File: app/src/main/java/com/haruhi/bismark439/haruhiism/TypeWriter.java
package com.haruhi.bismark439.haruhiism;

import android.content.Context;
import android.os.Handler;
import android.util.AttributeSet;

/**
 * Created by Bismark439 on 02/03/2018.
 */


public class TypeWriter extends android.support.v7.widget.AppCompatTextView {
    //Main Strings
    private String[]mStrings;
    private String stringSoFar;
    private int arrayIndex;

    //Sub Strings
    private String subString;
    private int mIndex;
    private long mDelay = 150; //Default 150ms delay

    //Beeping
    private boolean beep = true;
    private int repeat = 50;

    public TypeWriter(Context context) {
        super(context);
    }

    public TypeWriter(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    private Handler mHandler = new Handler();
    private Runnable characterAdder = new Runnable() {
        @Override
        public void run() {
            if(mStrings[arrayIndex].equals("YUKI.N>")){
                stringSoFar = appendText();
                setText(stringSoFar+"YUKI.N>_");
                arrayIndex++;
                if(arrayIndex < mStrings.length) {
                    mHandler.postDelayed(characterAdder, mDelay);
                }
            }else if(mStrings[arrayIndex].equals("\n")){//New line mode
                stringSoFar = appendText();
                setText(stringSoFar);
                arrayIndex++;
                mDelay = 780;
                if(arrayIndex < mStrings.length) {
                    mHandler.postDelayed(characterAdder, mDelay);
                }
            } else{//SubString Mode
                if(subString==null){
                    subString = mStrings[arrayIndex];
                    stringSoFar = appendText();
                    if(subString.length()>10){
                        mDelay=90;
                    }else{
                        mDelay=150;
                    }
                }
                if(mIndex==0){
                    setText(stringSoFar+subString.substring(0,mIndex));
                }else{
                    setText(stringSoFar+subString.substring(0,mIndex)+"_");
                }
                mIndex++;
                if(mIndex <= subString.length()) {
                    mHandler.postDelayed(characterAdder, mDelay);
                }else{
                    arrayIndex++;
                    mIndex=0;
                    subString=null;
                    if(arrayIndex < mStrings.length) {
                        mHandler.postDelayed(characterAdder, mDelay);
                    }else { //Wrote Everything
                        System.out.println("Send beeper");
                        stringSoFar = appendText();
                        mDelay=300;
                        repeat=500;
                        mHandler.postDelayed(beeper, mDelay);
                    }
                }
            }


        }
    };
    private Runnable beeper = new Runnable() {
        @Override
        public void run() {
            if(repeat>0){
                if(beep){
                    setText(stringSoFar);
                }else{
                    setText(stringSoFar+"_");
                }
                beep=!beep;
                repeat--;
                mHandler.postDelayed(beeper, mDelay);
            }
           // System.out.println(repeat);
        }
    };
    public void animateText(CharSequence text) {
        mIndex = 0;
        arrayIndex = 0;
        repeat = 0;
        mStrings = text.toString().split(",");
        setText("");
        mHandler.removeCallbacks(characterAdder);
        mHandler.postDelayed(characterAdder, mDelay);
    }
    public String appendText(){
        String text = "";
        for(int i=0;i<arrayIndex;i++){
            text=text+mStrings[i];
        }
        return text;
    }
    public void setCharacterDelay(long millis) {
        mDelay = millis;
    }
}

File: app/src/main/java/com/haruhi/bismark439/haruhiism/views/TypeWriter.kt
package com.haruhi.bismark439.haruhiism.views

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import com.haruhi.bismark439.haruhiism.R

/**
 * Created by Bismark439 on 02/03/2018.
 */
class TypeWriterView : AppCompatTextView {
    //Main Strings
    private var mStrings: Array<String> = arrayOf()
    private var stringSoFar: String? = null
    private var arrayIndex = 0

    //Sub Strings
    private var subString: String? = null
    private var mIndex = 0
    private var mDelay: Long = 150 //Default 150ms delay

    //Beeping
    private var beep = true
    private var repeat = 50

    constructor(context: Context?) : super(context!!)
    constructor(context: Context?, attrs: AttributeSet?) : super(context!!, attrs)

    private val mHandler = Handler(Looper.myLooper()!!)
    private val characterAdder: Runnable = object : Runnable {
        override fun run() {
            if (mStrings[arrayIndex] == "YUKI.N>") {
                stringSoFar = appendText()
                text = stringSoFar + context.getString(R.string.typewriter_yuki_n)
                arrayIndex++
                if (arrayIndex < mStrings.size) {
                    mHandler.postDelayed(this, mDelay)
                }
            } else if (mStrings[arrayIndex] == "\n") { //New line mode
                stringSoFar = appendText()
                text = stringSoFar
                arrayIndex++
                mDelay = 780
                if (arrayIndex < mStrings.size) {
                    mHandler.postDelayed(this, mDelay)
                }
            } else { //SubString Mode
                if (subString == null) {
                    subString = mStrings[arrayIndex]
                    stringSoFar = appendText()
                    mDelay = if (subString!!.length > 10) {
                        90
                    } else {
                        150
                    }
                }
                text = if (mIndex == 0) {
                    stringSoFar + subString!!.substring(0, mIndex)
                } else {
                    stringSoFar + subString!!.substring(0, mIndex) + "_"
                }
                mIndex++
                if (mIndex <= subString!!.length) {
                    mHandler.postDelayed(this, mDelay)
                } else {
                    arrayIndex++
                    mIndex = 0
                    subString = null
                    if (arrayIndex < mStrings.size) {
                        mHandler.postDelayed(this, mDelay)
                    } else { //Wrote Everything
                        println("Send beeper")
                        stringSoFar = appendText()
                        mDelay = 300
                        repeat = 500
                        mHandler.postDelayed(beeper, mDelay)
                    }
                }
            }
        }
    }
    private val beeper: Runnable = object : Runnable {
        override fun run() {
            if (repeat > 0) {
                text = if (beep) {
                    stringSoFar
                } else {
                    stringSoFar + "_"
                }
                beep = !beep
                repeat--
                mHandler.postDelayed(this, mDelay)
            }
            // System.out.println(repeat);
        }
    }

    fun animateText(text: CharSequence) {
        mIndex = 0
        arrayIndex = 0
        repeat = 0
        mStrings = text.toString().split(",").toTypedArray()
        setText("")
        mHandler.removeCallbacks(characterAdder)
        mHandler.postDelayed(characterAdder, mDelay)
    }

    fun appendText(): String {
        var text = ""
        for (i in 0 until arrayIndex) {
            text += mStrings[i]
        }
        return text
    }

    fun setCharacterDelay(millis: Long) {
        mDelay = millis
    }
}